### PR TITLE
feat(semantics): add DeclarationWalker for centralized block traversal and map key completion

### PR DESCRIPTION
**Pull request by albertocavalcante**
# Description

Centralizes duplicated block traversal logic into a shared `DeclarationWalker` helper and adds map literal key completion support.

### Phase 1: DeclarationWalker Refactor

- **New**: `DeclarationWalker.kt` - Shared helper for walking blocks and extracting declarations
  - Traverses all control flow: `if/for/while/do-while/try-catch/switch/synchronized`
  - Captures map literal keys and their value types for completion
- **Updated**: `GroovySemantics.populateScopeFromBlock` - Uses DeclarationWalker
- **Updated**: `SymbolExtractor.extractVariableSymbols` - Uses DeclarationWalker
- **New**: `DeclarationWalkerTest.kt` - 19 QA-focused unit tests

### Phase 2: Map Key Completion

- **Updated**: `CompletionProvider.handleMemberAccessContext` - Adds Strategy 0 for map literal keys
- **New helpers**: `findEnclosingBlock`, `addMapLiteralKeyCompletions`
- **New tests**: 5 completion scenarios (script-level, class method, if-block, empty map)

## Example

```groovy
def config = [host: "localhost", port: 8080]
config.  // Now suggests: host, port (plus standard map methods)
```

# Type of Change

- [ ] Bug fix
- [x] New feature
- [x] Refactoring

# Checklist

- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

# Verification

- `./gradlew :semantics-native:test` - 19 DeclarationWalker tests pass
- `./gradlew :groovy-lsp:test --tests "*.CompletionProviderTest"` - 15 tests pass

---

> [!NOTE]
> Unifies declaration traversal and enables smarter completions.
> 
> - **New**: `DeclarationWalker` centralizes walking `BlockStatement`s (if/for/while/do-while/try-catch/switch/synchronized), extracts variable types, and optionally captures map literal keys
> - **Semantics/Extraction**: `GroovySemantics.populateScopeFromBlock` and `SymbolExtractor.extractVariableSymbols` now use `DeclarationWalker`; adds `GroovySemantics.getContext`
> - **Completions**: `CompletionProvider` adds Strategy 0 for map-literal key suggestions (via `findEnclosingBlock` and `addMapLiteralKeyCompletions`) alongside normal map methods; minor type/import cleanups
> - **Tests**: New `DeclarationWalkerTest` (broad QA coverage) and expanded `CompletionProviderTest` (map keys across script/method/blocks, empty map); generics parsing in signatures verified
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7fd6c67620cea015eb00d2132b051b7273a59354. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Map literal key completion now suggests available keys contextually (script, methods, blocks) and shows inferred value types alongside standard map methods.

* **Bug Fixes / Stability**
  * Empty maps handled gracefully; unrelated keywords are no longer suggested.

* **Tests**
  * Expanded test coverage for map-key suggestions across scopes, blocks, and edge cases to ensure robustness.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: semantics/native/src/test/kotlin/com/github/albertocavalcante/gvy/semantics/native/DeclarationWalkerTest.kt [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+                }
+            """.trimIndent()
+
+            val module = parse(code)
+            val context = getContext(module)
+            val block = getMethodBlock(module, "test")
+
+            val result = DeclarationWalker.walk(block, context)
+
+            // This tests a GAP - the current implementation doesn't traverse IfStatement blocks!
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply semantics/native/src/test/kotlin/com/github/albertocavalcante/gvy/semantics/native/DeclarationWalkerTest.kt]: Fixed in 07f23718. Updated comment to reflect that if-block traversal gap is now fixed.

File: semantics/native/src/test/kotlin/com/github/albertocavalcante/gvy/semantics/native/DeclarationWalkerTest.kt [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+            val context = getContext(module)
+            val block = getMethodBlock(module, "test")
+
+            val result = DeclarationWalker.walk(block, context)
+
+            // For loop variable 'i' and body variable 'loopVar' should both be captured
+            assertTrue(
+                result.variables.any { it.name == "loopVar" },
+                "Declaration inside for loop body should be captured",
+            )
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply semantics/native/src/test/kotlin/com/github/albertocavalcante/gvy/semantics/native/DeclarationWalkerTest.kt]: Fixed in 07f23718. Updated test comment to accurately reflect what is tested.

File: semantics/native/src/main/kotlin/com/github/albertocavalcante/gvy/semantics/native/SymbolExtractor.kt [hunk 3]
```diff
@@ -52,50 +52,70 @@ object SymbolExtractor {
...
-                    }
+            val context = if (semantics != null && module != null) {
+                semantics.getContext(module)
+            } else {
+                null
+            }
 
+            if (context != null) {
+                // Use DeclarationWalker for proper type inference
+                val result = DeclarationWalker.walk(code, context, captureMapKeys = true)
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply semantics/native/src/main/kotlin/com/github/albertocavalcante/gvy/semantics/native/SymbolExtractor.kt]: Fixed in 07f23718. Set captureMapKeys=false since mapKeys is not used in VariableSymbol.

File: groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProvider.kt [hunk 4]
```diff
@@ -520,6 +535,60 @@ object CompletionProvider {
...
+        ctx: CompletionContext,
+    ): Boolean {
+        val moduleNode = ctx.moduleNode ?: return false
+        val nativeContext = ctx.semanticResolver.semantics.getContext(moduleNode)
+            ?: return false
+
+        val block = findEnclosingBlock(ctx) ?: return false
+        val result = DeclarationWalker.walk(block, nativeContext, captureMapKeys = true)
+
+        val mapDecl = result.variables.find { it.name == qualifierName }
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProvider.kt]: Fixed in 42623cc4. Updated to use findLast to correctly handle shadowed variables by picking the innermost declaration.

File: semantics/native/src/main/kotlin/com/github/albertocavalcante/gvy/semantics/native/DeclarationWalker.kt [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+                }
+            }
+
+            is BlockStatement -> walkBlock(stmt, context, captureMapKeys, out)
+            is IfStatement -> {
+                walkChild(stmt.ifBlock, context, captureMapKeys, out)
+                walkChild(stmt.elseBlock, context, captureMapKeys, out)
+            }
+
+            is ForStatement -> walkChild(stmt.loopBlock, context, captureMapKeys, out)
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply semantics/native/src/main/kotlin/com/github/albertocavalcante/gvy/semantics/native/DeclarationWalker.kt]: Fixed in 42623cc4. Added logic to capture the ForStatement variable (e.g., 'i' in 'for(int i=0;...)') and variable from ClosureListExpression (C-style loops).

File: groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProvider.kt [hunk 6]
```diff
@@ -520,6 +536,60 @@ object CompletionProvider {
...
+        val classNode = moduleNode.classes.firstOrNull()
+        if (classNode != null) {
+            val method = classNode.methods.find { method ->
+                method.lineNumber <= ctx.line + 1 &&
+                    (method.lastLineNumber >= ctx.line + 1 || method.lastLineNumber <= 0)
+            }
+            if (method?.code is BlockStatement) {
+                return method.code as BlockStatement
+            }
+        }
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProvider.kt]: Fixed in 42623cc4. findEnclosingBlock now iterates all classes in the module.

File: groovy-lsp/src/test/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProviderTest.kt [hunk 7]
```diff
@@ -179,42 +179,184 @@ class CompletionProviderTest {
...
+        assertTrue(
+            completions.any { it.label == "port" },
+            "Should suggest 'port' map key. Found: ${completions.map { it.label }}",
+        )
 
-        // Assert - Should NOT have keywords (proves context detection works)
+        // Should NOT have keywords
         assertTrue(
             completions.none { it.label == "abstract" },
             "Should NOT suggest 'abstract' keyword for map member access",
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-lsp/src/test/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProviderTest.kt]: Fixed in 42623cc4. Clarified test assertion message.

File: groovy-lsp/src/test/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProviderTest.kt [hunk 8]
```diff
@@ -179,42 +179,184 @@ class CompletionProviderTest {
...
+
     @Test
-    fun `should suggest map methods for map literal variable - unmocked`() = runTest {
-        // Arrange - use real SemanticTypeResolver with a stub TypeSolver
+    fun `should suggest map literal keys for script-level declaration`() = runTest {
         val stubTypeSolver = object : TypeSolver {
             override var parent: TypeSolver? = null
             override fun tryToSolveType(name: String) = SymbolReference.unsolved<ResolvedTypeDeclaration>()
         }
         val realSemanticResolver = SemanticTypeResolver(stubTypeSolver)
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-lsp/src/test/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProviderTest.kt]: Deferring to #841. Will refactor test duplication in a follow-up.

File: groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProvider.kt [hunk 9]
```diff
@@ -520,6 +535,60 @@ object CompletionProvider {
...
+     */
+    private fun findEnclosingBlock(ctx: CompletionContext): BlockStatement? {
+        val moduleNode = ctx.moduleNode ?: return null
+
+        // Check if cursor is inside a class method
+        val classNode = moduleNode.classes.firstOrNull()
+        if (classNode != null) {
+            val method = classNode.methods.find { method ->
+                method.lineNumber <= ctx.line + 1 &&
+                    (method.lastLineNumber >= ctx.line + 1 || method.lastLineNumber <= 0)
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProvider.kt]: Deferring clarification/refinement of line number logic to #841. The current logic handles missing end lines optimistically.

File: semantics/native/src/main/kotlin/com/github/albertocavalcante/gvy/semantics/native/SymbolExtractor.kt [hunk 10]
```diff
@@ -52,50 +52,70 @@ object SymbolExtractor {
...
+                                line = stmt.lineNumber - 1,
+                            ),
+                        )
+                    }
+                }
+
+                is BlockStatement -> extractVariablesWithoutSemantics(stmt, out)
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply semantics/native/src/main/kotlin/com/github/albertocavalcante/gvy/semantics/native/SymbolExtractor.kt]: Deferring SymbolExtractor alignment to #841.

File: groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProvider.kt [hunk 11]
```diff
@@ -520,6 +536,59 @@ object CompletionProvider {
...
+
+        // Check if cursor is inside any class method
+        for (classNode in moduleNode.classes) {
+            val method = classNode.methods.find { method ->
+                method.lineNumber <= ctx.line + 1 &&
+                    (method.lastLineNumber >= ctx.line + 1 || method.lastLineNumber <= 0)
+            }
+            if (method?.code is BlockStatement) {
+                return method.code as BlockStatement
+            }
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/providers/completion/CompletionProvider.kt]: Addressed in 7fd6c676 by using findLast to pick the innermost method scope. Deferring further refinement (minByOrNull) to #841.
